### PR TITLE
feat: Create metadata DTOs and models for V2 API

### PR DESCRIPTION
**Pull request by weichou1229**
Create metadata DTOs and models for Addressable, Device, DeviceProfile, DeviceService.
Fix: #249

Signed-off-by: weichou <weichou1229@gmail.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://wiki.edgexfoundry.org/display/FA/Contributor%27s+Guide
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number:

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

## Other information

### CODE REVIEW

File: v2/dtos/autoevent.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	// It represents as a duration string.
+	// The format of this field is to be an unsigned integer followed by a unit which may be "ms", "s", "m" or "h"
+	// representing milliseconds, seconds, minutes or hours. Eg, "100ms", "24h"
+	Frequency string `json:"frequency,omitempty"`
+	// OnChange indicates whether the device service will generate an event only,
+	// if the reading value is different from the previous one.
+	// If true, only generate events when readings change
+	OnChange bool `json:"onChange,omitempty"`
+	// Resource indicates the name of the resource in the device profile which describes the event to generate
+	Resource string `json:"resource,omitempty"`
```
Comment (cloudxxx8) [inline v2/dtos/autoevent.go:24]: this field is required

File: v2/dtos/autoevent.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+)
+
+// AutoEvent supports auto-generated events sourced from a device service
+type AutoEvent struct {
+	// Frequency indicates how often the specific resource needs to be polled.
+	// It represents as a duration string.
+	// The format of this field is to be an unsigned integer followed by a unit which may be "ms", "s", "m" or "h"
+	// representing milliseconds, seconds, minutes or hours. Eg, "100ms", "24h"
+	Frequency string `json:"frequency,omitempty"`
```
Comment (cloudxxx8) [inline v2/dtos/autoevent.go:18]: this field is required

File: v2/dtos/requests/device.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	ProfileName        string                             `json:"profileName" validate:"required"`
+	AdminState         string                             `json:"adminState"`
+	OperatingState     string                             `json:"operatingState"`
+	AutoEvents         []dtos.AutoEvent                   `json:"autoEvents"`
+	Protocols          map[string]dtos.ProtocolProperties `json:"protocols" validate:"required,gt=0"`
+}
+
+// Validate satisfies the Validator interface
+func (d AddDeviceRequest) Validate() error {
+	err := v2.Validate(d)
```
Comment (cloudxxx8) [inline v2/dtos/requests/device.go:33]: add logic to validate AutoEvent.Frequency format

File: v2/dtos/requests/device.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	OperatingState     string                             `json:"operatingState"`
+	AutoEvents         []dtos.AutoEvent                   `json:"autoEvents"`
+	Protocols          map[string]dtos.ProtocolProperties `json:"protocols"`
+	LastConnected      int64                              `json:"lastConnected"`
+	LastReported       int64                              `json:"lastReported"`
+	Notify             bool                               `json:"notify"`
+}
+
+// Validate satisfies the Validator interface
+func (dp UpdateDeviceRequest) Validate() error {
```
Comment (cloudxxx8) [inline v2/dtos/requests/device.go:115]: add logic to validate AutoEvent.Frequency format

File: v2/dtos/requests/device.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		d.Protocols = ProtocolDTOsToModels(req.Protocols)
+		Devices = append(Devices, d)
+	}
+	return Devices
+}
+
+// UpdateDeviceRequest defines the Request Content for PUT event as pushed DTO.
+// This object and its properties correspond to the UpdateDeviceRequest object in the APIv2 specification:
+// https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-metadata/2.x#/UpdateDeviceRequest
+type UpdateDeviceRequest struct {
```
[the remark this answers is not in the document]
Comment (weichou1229) [reply v2/dtos/requests/device.go:99]: Use `required_without ` to validate the field:
* require the Id if the DeviceName is not present
* require the DeviceName if the Id is not present

https://pkg.go.dev/github.com/go-playground/validator/v10?tab=doc#hdr-Required_Without

File: v2/dtos/requests/addressable.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+// This object and its properties correspond to the UpdateAddressableRequest object in the APIv2 specification:
+// https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-metadata/2.x#/UpdateAddressableRequest
+type UpdateAddressableRequest struct {
+	common.BaseRequest `json:",inline"`
+	Id                 string `json:"id" validate:"required"`
+	Name               string `json:"name" validate:"required"`
+	Protocol           string `json:"protocol,omitempty"`
+	Method             string `json:"method,omitempty"`
+	Address            string `json:"address,omitempty"`
+	Port               int16  `json:"port,omitempty,Number"`
```
[the remark this answers is not in the document]
Comment (weichou1229) [reply v2/dtos/requests/addressable.go:96]: Change to uint16,  add testing to validate the invalid port number and JSON unmarshal function can throw the error:
`Failed to unmarshal request body as JSON, json: cannot unmarshal number 4294967295 into Go struct field .Port of type uint16.`

File: v2/dtos/autoevent.go (hunk 1 above)
Comment (weichou1229) [reply v2/dtos/autoevent.go]: Modified.

File: v2/dtos/autoevent.go (hunk 2 above)
Comment (weichou1229) [reply v2/dtos/autoevent.go]: Modified.

File: v2/dtos/requests/device.go (hunk 3 above)
Comment (weichou1229) [reply v2/dtos/requests/device.go:37]: Added.

File: v2/dtos/requests/device.go (hunk 4 above)
Comment (weichou1229) [reply v2/dtos/requests/device.go]: Added.

File: v2/dtos/requests/addressable.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (cloudxxx8) [reply v2/dtos/requests/addressable.go]: is this failure resolved?

[the remark this answers is not in the document]
Comment (weichou1229) [reply v2/dtos/requests/addressable.go]: Yes.

**Pull request merged**